base/show.jl

```diff
@@ -922,13 +922,22 @@ function ismodulecall(ex::Expr)
 end
 
 function show(io::IO, tv::TypeVar)
+    # If `tvar_env` exist and we are in it, the type constraint are
+    # already printed and we don't need to print it again.
+    # Otherwise, the lower bound should be printed if it is not `Bottom`
+    # and the upper bound should be printed if it is not `Any`.
+    # The upper bound `Any` should also be printed if we are not in the
+    # existing `tvar_env` in order to resolve the ambiguity when printing a
+    # method signature.
+    # i.e. `foo{T,N}(::Array{T,N}, ::Array)` should be printed as
+    # `foo{T,N}(::Array{T,N}, ::Array{T<:Any,N<:Any})`
```

Comment:
I kind of agree... This comment is just describing the intent of https://github.com/JuliaLang/julia/commit/fff2b9aa811549186ec39e71cc4f6331203e2119 so that I can figure out what is the right fix.

To get a less verbose printing, should we print `Array` as `Array` in all cases (currently it prints as `Array{T,N}`) or only for method signatures?

Also what about types with partially specified parameters? How should `::Vector` be printed (I don't think we store the name of the `TypeConstructor` so we have to print it as `Array{...}`)?